pennylane/fermi/fermionic.py

```diff
@@ -303,6 +303,132 @@ def to_mat(self, n_orbitals=None):
+
+        Args:
+            source (int): The position of the operator to be commuted.
+            target (int): The desired position of the operator occupying the source position.
+
+        Returns:
+            FermiSentence: The ``FermiSentence`` obtained after applying the anti-commutator relations.
+
+        Raises:
+            TypeError: If the source or target is not an integer
+            ValueError: If the source or target are outside the range [0, len(self) - 1]
+
+
+        **Example**
+
+        >>> w = FermiWord({(0, 0): '+', (1, 1): '-'})
+        >>> w.commute(0, 1)
+        -1 * a(1) a⁺(0)
+        """
+
+        if not isinstance(source, int) or not isinstance(target, int):
+            raise TypeError("Indices must be integers")
+
+        if source < 0 or target < 0:
+            raise ValueError("Indices must be positive integers")
+
+        if source > len(self.sorted_dic) - 1 or target > len(self.sorted_dic) - 1:
+            raise ValueError("Indices out of range")
+
+        if source == target:
+            return FermiSentence({self: 1})
+
+        fw = self
+        fs = FermiSentence({self: 1})
+        delta = 1 if source < target else -1
+
+        while source != target:
+            fs, fw = _commute_adjacent(fs, fw, source, source + delta)
+            source += delta
+
+        return fs
+
+
+def _commute_adjacent(fs, fw, i, j):
+    """Commutes two adjacent operators within a ``FermiSentence``.
+
+    Args:
+        fs (``FermiSentence``): The ``FermiSentence`` to be modified
+        fw (``FermiWord``): The ``FermiWord`` within ``fs`` that contains the operators
+            to be commuted.
+        source (int): The position of the operator to be commuted.
+        target (int): The desired position of the operator occupying the source position.
+
+    Returns:
+        FermiSentence: The ``FermiSentence`` obtained after applying the anti-commutator relations.
+
+    Raises:
+        ValueError: If the source and target positions are not consecutive
+    """
+    if i != j + 1 and j != i + 1:
+        raise ValueError("Indices must be consecutive integers")
+
+    indices = list(fw.sorted_dic.keys())
+    i_idx = indices[i]
+    j_idx = indices[j]
+
+    i_val = fw[i_idx]
+    j_val = fw[j_idx]
+
+    # commuting identical terms
+    if i_idx[1] == j_idx[1] and i_val == j_val:
+        return fs, fw
+
+    new_i_idx = (i, j_idx[1])
+    new_j_idx = (j, i_idx[1])
+
+    coeff = fs[fw]
+    del fs[fw]
+    fw = dict(fw)
+    fw[new_i_idx] = j_val
+    fw[new_j_idx] = i_val
+
+    if i_idx[1] != j_idx[1]:
+        del fw[i_idx]
+        del fw[j_idx]
+
+    fw = FermiWord(fw)
+
+    fs = dict(fs)
+    fs = FermiSentence(fs)
+
+    if i_val == j_val or i_idx[1] != j_idx[1]:
+        fs = fs + (-1 * FermiSentence({fw: coeff}))
+    else:
+        left = {}
+        lpos = 0
+
+        middle = {}
+        mpos = 0
+
+        right = {}
+        rpos = 0
+
+        for key, value in fw.sorted_dic.items():
```

Comment:
Could we make it work based on indexing instead? 🤔

Reply:
I rewrote it based on index slicing. Let me know if you had something else in mind.